WooCommerce/Classes/POS/ViewModels/AnyTotalsViewModel.swift

```diff
@@ -0,0 +1,138 @@
+import Combine
+import Yosemite
+
+class AnyTotalsViewModel: ObservableObject, TotalsViewModelProtocol {
```

Comment:
Thanks for your review. 

> it also shows how TotalsViewModel, CardViewModel, and PointOfSaleDashboardViewModel are all referencing each other

> I forgot about the complication of using ObservableObject as a protocol. 

Indeed. Since taking on this work I've struggled with tradeoffs.

My current set of changes which properly mock dependencies produce a larger diff and have other downstream effects as you've noted than my earlier approach to adding test coverage which were a pragmatic solution but required undesirable changes to the class under test. 

In the latest approach `PointOfSaleDashboardViewModel` is in better shape, but some complexity cropped up while abstracting its dependencies.

Regarding

> shows how TotalsViewModel, CardViewModel, and PointOfSaleDashboardViewModel are all referencing each other.

This goes to the heart of the matter. It look like `PointOfSaleDashboardViewModel` wasn't designed with testing in mind. It had tight coupling between `PointOfSaleDashboardViewModel`, `CartViewModel`, and `TotalsViewModel` which I've reduced by introducing abstractions including interface-based interactions, type erasure and dependency injection.

>review the dependency graph at the end of M2 to make sure it's all necessary

I too wonder if a more fundamental refactoring is in order at some point.

In the meantime though I'll take a look at your suggestion of using generics. Thanks for that.

Reply:
Yes, I understand the complexities now after touching the code. The setup is hard to test which shows problems with decoupling. We tried to decouple the logic by splitting it into different classes or files but it all still remains coupled.

> In the meantime though I'll take a look at your suggestion of using generics. Thanks for that.

Okay, unfortunately, it also introduces a bunch of cascading changes due to same reasons. However, if we continue using observable VMs within other VMs, then generics may be the only way that allows us to inject mocks cleanly.

Reply:
Thanks again for your thorough review and draft solution using generics. 

I agree that using generics could help inject mocks cleanly and address the issue of assigning publishers in initializers. However, this approach also introduces a layer of complexity and, as you noted, a different set of cascading changes, which might slow down our progress in the short term.

After evaluating the pros and cons I think my current approach, with improvements to address your concerns about manual publisher initialization in `AnyTotalsViewModel`, is preferable to using generics to accomplish the decoupling needed to expand test coverage.


Here's my rationale.

1. Improved publisher initialization
Commit b14e014047698dbdfaf7df70543b62514d9124f3 responds to your feedback about manual publisher initialization. 

2. Speed of development: Changes from `trunk` have been merged in. Types are more loosely coupled and test coverage of `PointOfSaleDashboardViewModel` expanded using protocols. These make the current implementation a move in the right direction.

3. Improved readability and lower overall complexity: I find the protocol-based code more readable and easier to understand. It doesn't have the syntax verbosity that comes with generics. 
4. Precedence in the codebase. It looks like there are limited example in the codebase of [using generics for view models](https://github.com/woocommerce/woocommerce-ios/blob/10900a66bb13b3c611590d66acae6cbcc4c72858/WooCommerce/WooCommerceTests/POS/ViewModels/PointOfSaleDashboardViewModelTests.swift#L23). Just one I think based on a search of `ViewModel<`. However searching `ViewModel(` yields many results.

Regarding:
>However, if we continue using observable VMs within other VMs, then generics may be the only way that allows us to inject mocks cleanly.

This isn't a pattern I would recommend we continue to follow. Being the exception rather than the rule can we reserve generics for this situation if it continues to arise?

Reply:
Thanks. Yes, I agree with you that in the context of `generics` vs `type-erased AnyVM`, the latter has sound arguments. 

However, by looking at the solution again, it still feels like we're trying to go against the system frameworks and the language. I want to avoid us choosing an overcomplicated solution and slowing down our M2 work.

We want to keep things testable but also we want keep things simple.

---

I'll try to go through the problem again.

#### What is the problem?

We want to make our ViewModel code testable.

#### What is the common solution?

Use protocols and dependency injection.

#### What stopping us from using it?

Using protocols and `@ObservedObject` is not supported.

#### When it happens?

For example, in `PointOfSaleDashboardView` we're referencing `TotalsViewModel`. But we also want to inject `TotalsViewModel` into `PointOfSaleDashboardViewModel` for testability.

And I think that's the main problem. `SwiftUI` expects `View Model` to be used by `View`. However, we want to use `ViewModel` within `Views` AND within other `ViewModels`.

#### What can we do about it?

Let's use a concrete implementation in `Views`, and protocols in `View Models`. 

It's a short-term move that doesn't force us into major refactorings, using generics, or opting for `AnyViewModel` type of solutions.

The side effect is that it forces us to clean up dependency management and to make sure all of the view models are created in a single place. It transparently exposes the fact that our code is not decoupled but nothing will stop us from refactoring in the future.

Here's a WIP:

https://github.com/woocommerce/woocommerce-ios/compare/feat/13127-disable-cta-to-go-back-to-product-selector-when-order-is-being-synced-add-test...feat/13127-disable-cta-to-go-back-to-product-selector-when-order-is-being-synced-add-test-rearchitect?expand=1

I didn't update tests but I think it would work. The main changes are:

- Creating dependencies at entry point
- Passing dependencies directly to views and view models

It allows to expect protocols in VMs and concrete implementation in Views.

Let me know what you think.